pkg/pushsync/pushsync.go

```diff
@@ -195,25 +184,89 @@ func (ps *PushSync) chunksWorker() {
 	}
 }
 
-// sendChunkMsg sends a chunk to a given peer.
-func (ps *PushSync) sendChunkMsg(ctx context.Context, peer swarm.Address, ch swarm.Chunk) error {
+// ReceiveChunkAndSendReceipt receives a chunk delivery from a peer, stores it in the localDB and sends a receipt
+// in the same stream. Swarm book 2.3.2.
+func (ps *PushSync) ReceiveChunkAndSendReceipt(ctx context.Context, stream p2p.Stream) (swarm.Chunk, error) {
+	w, r := protobuf.NewWriterAndReader(stream)
+	defer stream.Close()
+
+	var ch pb.Delivery
+	if err := r.ReadMsg(&ch); err != nil {
+		ps.metrics.ReceivedChunkErrorCounter.Inc()
+		return  nil, err
+	}
+	ps.metrics.ChunksSentCounter.Inc()
+
+	// create chunk and store it in the local store
+	addr := swarm.NewAddress(ch.Address)
+	chunk := swarm.NewChunk(addr, ch.Data)
+	_, err := ps.storer.Put(ctx, storage.ModePutSync, chunk)
+	if err != nil {
+		return nil, err
+	}
+	ps.metrics.TotalChunksStoredInDB.Inc()
+
+	// Send a receipt immediately once the storage of the chunk is successfull
```

Comment:
this should happen  only if the node is closest to chunk, otherwise you need to wait for the receipt on the stream and  pass it back to the requestor

Reply:
it is...  As in the code now  we will et the delivery ONLY to the closest node.. the is no kademlia routing....
WHen we implement routing.. what you say is true.

Reply:
Yes... Now the chunk will reach only of the node is closest. Because we have not yet implemented kademlia. SO if you receive a chunk delivery, it means you are the closest node.. See function chunksWorker which delivers chunk only to closestpeers.

AFter kademlia is done... what you are saying is right.

Reply:
Since the first comment is accepted, this comment was also taken care of in the fix.